### PR TITLE
feat(Table): Editable cell

### PR DESCRIPTION
**Pull request by bentleyvk**
- Moved all cell related stuff to `TableCell`. 
- Callback is only called when `onBlur` is fired to avoid many `Table` rerenders when typing that causes refocus and caret position reset.
 - Made `isCellEditable` a function if someone might need conditionally enable edit. Just a safe bet.

editable-cell3

## Checklist

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] Approve test images for new stories
- [x] Add screenshots of the key elements of the component

### CODE REVIEW

File: src/core/Table/EditableCell.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+export type OnCellEditCallbackParams<T extends Record<string, unknown>> = {
+  columnId: string;
+  value: string;
+  rowData: T;
+};
+
+export type EditableCellProps<T extends Record<string, unknown>> = {
+  cell: Cell<T>;
+  cellProps: TableCellProps;
+  onCellEdit?: (cellEditParams: OnCellEditCallbackParams<T>) => void;
```
Comment (bentleyvk) [inline src/core/Table/EditableCell.tsx:17]: I am not sure whether to make params object or just each param separately `(columnId, value, rowData) => void`.
Currently made it as a object because I saw it somewhere like that 😀

**Comment by bentleyvk** [comment]
Visual tests should not fail after https://github.com/iTwin/iTwinUI-react/pull/306 is merged.

File: src/core/Table/TableCell.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import { getCellStyle } from './utils';
+import EditableCell, { OnCellEditCallbackParams } from './EditableCell';
+import SubRowExpander from './SubRowExpander';
+import { SELECTION_CELL_ID } from './hooks';
+
+export type TableCellProps<T extends Record<string, unknown>> = {
+  cell: Cell<T>;
+  cellIndex: number;
+  onCellEdit?: (cellEditParams: OnCellEditCallbackParams<T>) => void;
+  isRowDisabled: boolean;
```
Comment (mayank99) [inline src/core/Table/TableCell.tsx:17]: `isDisabled` would make more sense, no? Why do we care about row?

File: src/core/Table/TableCell.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  tableHasSubRows: boolean;
+  tableInstance: TableInstance<T>;
+  expanderCell?: (cellProps: CellProps<T>) => React.ReactNode;
+};
+
+const TableCell = <T extends Record<string, unknown>>(
```
Comment (mayank99) [inline src/core/Table/TableCell.tsx:23]: export this?

File: src/core/index.ts [hunk 4]
```diff
@@ -122,6 +122,7 @@ export type {
   TableFilterValue,
   DateRangeFilterOptions,
   FilterButtonBarProps,
+  OnCellEditCallbackParams,
```
Comment (mayank99) [inline src/core/index.ts:125]: Not sure we need to export this.

File: src/core/Table/EditableCell.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  const [isDirty, setIsDirty] = React.useState(false);
+
+  return (
+    <div
+      {...cellProps}
+      contentEditable
+      suppressContentEditableWarning
+      onInput={(e) => {
+        setValue(
+          (e.target as HTMLElement).innerText.replace(/(\r\n|\n|\r)/gm, ''),
```
Comment (mayank99) [inline src/core/Table/EditableCell.tsx:36]: Maybe add a comment saying you're removing new lines here

File: src/core/Table/EditableCell.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  children: React.ReactNode;
+};
+
+const EditableCell = <T extends Record<string, unknown>>(
+  props: EditableCellProps<T>,
+) => {
+  const { cell, cellProps, onCellEdit, children } = props;
+
+  const [value, setValue] = React.useState(cell.value);
+  const [isDirty, setIsDirty] = React.useState(false);
```
Comment (mayank99) [inline src/core/Table/EditableCell.tsx:27]: Is this a standard name? Something like `didChange` or `isChanged` or `isModified` might be more clear.

Or add a comment.

File: src/core/Table/TableCell.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import { SELECTION_CELL_ID } from './hooks';
+
+export type TableCellProps<T extends Record<string, unknown>> = {
+  cell: Cell<T>;
+  cellIndex: number;
+  onCellEdit?: (cellEditParams: OnCellEditCallbackParams<T>) => void;
+  isRowDisabled: boolean;
+  tableHasSubRows: boolean;
+  tableInstance: TableInstance<T>;
+  expanderCell?: (cellProps: CellProps<T>) => React.ReactNode;
```
Comment (mayank99) [inline src/core/Table/TableCell.tsx:20]: This is not really a "cell" if it's part of another cell.

File: src/core/Table/EditableCell.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+};
+
+export type EditableCellProps<T extends Record<string, unknown>> = {
+  cell: Cell<T>;
+  cellProps: TableCellProps;
+  onCellEdit?: (cellEditParams: OnCellEditCallbackParams<T>) => void;
+  children: React.ReactNode;
+};
+
+const EditableCell = <T extends Record<string, unknown>>(
```
Comment (mayank99) [inline src/core/Table/EditableCell.tsx:21]: export module

File: src/core/Table/EditableCell.tsx [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      suppressContentEditableWarning
+      onInput={(e) => {
+        setValue(
+          (e.target as HTMLElement).innerText.replace(/(\r\n|\n|\r)/gm, ''),
+        );
+        setIsDirty(true);
+      }}
+      onBlur={() =>
+        isDirty &&
+        onCellEdit?.({
```
Comment (mayank99) [inline src/core/Table/EditableCell.tsx:42]: So I was thinking that we might handle the cell editing internally, i.e. update table data on blur.

But that might be more complicated and less flexible, which makes me wonder... Why not just provide this `EditableCell` component (with styling) for the user to pass into `Cell` or `defaultCell`? Or is there some data/state that the user may not have access to?

File: src/core/Table/EditableCell.tsx (hunk 1 above)
Comment (mayank99) [reply src/core/Table/EditableCell.tsx:17]: object is less clear to me

File: src/core/Table/EditableCell.tsx (hunk 6 above)
Comment (bentleyvk) [reply src/core/Table/EditableCell.tsx:27]: I know Angular forms had this field `isDirty`

File: src/core/Table/TableCell.tsx (hunk 7 above)
Comment (bentleyvk) [reply src/core/Table/TableCell.tsx:20]: Just left the same naming as it comes from the `Table` 🤷‍♂️

File: src/core/Table/EditableCell.tsx (hunk 9 above)
Comment (mayank99) [reply src/core/Table/EditableCell.tsx:42]: Yeah, I think `cellRenderer` would be a better approach than adding all these extra props. It would also future proof us for other kinds of cells (e.g. status cell).

File: src/core/Table/EditableCell.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  props: EditableCellProps<T>,
+) => {
+  const { cell, cellProps, onCellEdit, children } = props;
+
+  const [value, setValue] = React.useState(cell.value);
+  const [isDirty, setIsDirty] = React.useState(false);
+
+  return (
+    <div
+      {...cellProps}
```
Comment (mayank99) [inline src/core/Table/EditableCell.tsx:31]: I think the user should have some way to override or add to `onInput` and `onBlur`.

The current approach of holding update until onBlur is a bit fragile. What if onBlur is never called because the end-user stays on the cell for too long? In such a case, it might be helpful to update the table after a set delay.

File: stories/core/Table.stories.tsx [hunk 11]
```diff
@@ -1324,3 +1325,100 @@ Full.args = {
...
+            isCellEditable: (rowData) => rowData.description !== 'Fetching...',
+            Filter: tableFilters.TextFilter(),
+          },
+        ],
+      },
+    ],
+    [],
+  );
+
+  const onCellEdit = useCallback(
```
Comment (mayank99) [inline stories/core/Table.stories.tsx:1372]: Not sure if you plan to keep this approach, but if you do, then you should add an example like this callback in the JSDoc comment of `onCellEdit` prop. Because it's not very obvious that the user needs to update the data on their end.

File: src/core/Table/Table.test.tsx [hunk 12]
```diff
@@ -1486,3 +1486,97 @@ it('should render sub-rows with custom expander', () => {
...
+          id: 'description',
+          Header: 'description',
+          accessor: 'description',
+        },
+        {
+          id: 'view',
+          Header: 'view',
+          Cell: () => {
+            return <span>View</span>;
+          },
```
Comment (mayank99) [inline src/core/Table/Table.test.tsx:1513]: Huh?

Comment (mayank99) [reply src/core/Table/Table.test.tsx:1513]: Oh ok I see you copied from other test and removed `onClick`. Maybe could remove this entire column as it doesn't do anything for this test.

File: src/core/Table/TableCell.tsx [hunk 13]
```diff
@@ -0,0 +1,87 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { Cell, CellProps, TableInstance } from 'react-table';
+import cx from 'classnames';
+import { getCellStyle } from './utils';
+import EditableCell, { OnCellEditCallbackParams } from './EditableCell';
+import SubRowExpander from './SubRowExpander';
```
Comment (mayank99) [inline src/core/Table/TableCell.tsx:10]: Forgot to mention in previous PR but probably should export ES module for `SubRowExpander` instead of using default export.

File: src/core/Table/EditableCell.tsx [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+import React from 'react';
+import { Cell, TableCellProps } from 'react-table';
+
+export type OnCellEditCallbackParams<T extends Record<string, unknown>> = {
+  columnId: string;
+  value: string;
+  rowData: T;
+};
+
+export type EditableCellProps<T extends Record<string, unknown>> = {
```
Comment (gretanausedaite) [inline src/core/Table/EditableCell.tsx:14]: Maybe some comments or description is missing.

Comment (bentleyvk) [reply src/core/Table/EditableCell.tsx]: Added

File: src/core/Table/Table.test.tsx (hunk 12 above)
Comment (bentleyvk) [reply src/core/Table/Table.test.tsx:1515]: `assertRowsData` expects that there will be `View` column, so I need to add it.

File: src/core/Table/TableCell.tsx (hunk 2 above)
Comment (bentleyvk) [reply src/core/Table/TableCell.tsx]: Renamed

Comment (bentleyvk) [reply src/core/Table/TableCell.tsx]: Changed to just params

File: src/core/Table/TableCell.tsx (hunk 3 above)
Comment (bentleyvk) [reply src/core/Table/TableCell.tsx]: Exported

File: src/core/index.ts (hunk 4 above)
Comment (bentleyvk) [reply src/core/index.ts]: Removed

File: src/core/Table/EditableCell.tsx (hunk 5 above)
Comment (bentleyvk) [reply src/core/Table/EditableCell.tsx]: Added

File: src/core/Table/EditableCell.tsx (hunk 8 above)
Comment (bentleyvk) [reply src/core/Table/EditableCell.tsx]: Exported

File: stories/core/Table.stories.tsx (hunk 11 above)
Comment (bentleyvk) [reply stories/core/Table.stories.tsx]: Added a comment on `onCellEdit` prop.

File: src/core/Table/TableCell.tsx (hunk 13 above)
Comment (bentleyvk) [reply src/core/Table/TableCell.tsx]: Added

File: src/core/Table/EditableCell.tsx (hunk 10 above)
Comment (bentleyvk) [reply src/core/Table/EditableCell.tsx]: Extended div props and calling user `onInput` and `onBlur` inside ours.

File: src/core/Table/TableCell.tsx [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+          isDisabled={isDisabled}
+          cellProps={cellProps}
+          expanderCell={expanderCell}
+        />
+      )}
+      {cell.render('Cell')}
+    </>
+  );
+
+  const cellRendererProps = {
```
Comment (veekeys) [inline src/core/Table/TableCell.tsx:76]: Maybe define the type?

File: src/core/Table/TableCell.tsx [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+export type TableCellProps<T extends Record<string, unknown>> = {
+  cell: Cell<T>;
+  cellIndex: number;
+  isDisabled: boolean;
+  tableHasSubRows: boolean;
+  tableInstance: TableInstance<T>;
+  expanderCell?: (cellProps: CellProps<T>) => React.ReactNode;
+};
+
+export const TableCell = <T extends Record<string, unknown>>(
```
Comment (veekeys) [inline src/core/Table/TableCell.tsx:22]: Why this is not in the cells folder?

File: src/core/Table/Table.test.tsx [hunk 17]
```diff
@@ -1486,3 +1487,51 @@ it('should render sub-rows with custom expander', () => {
...
+  const rows = container.querySelectorAll('.iui-table-body .iui-row');
+  assertRowsData(rows);
+
+  const editableCells = container.querySelectorAll(
+    '.iui-cell[contenteditable]',
+  );
+  expect(editableCells).toHaveLength(3);
+
+  fireEvent.input(editableCells[1], {
+    target: { innerText: 'test-data\n\r\r\n' },
```
Comment (veekeys) [inline src/core/Table/Table.test.tsx:1533]: why new line is needed in inner text?

File: src/core/Table/cells/DefaultCell.tsx [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+import { CellRendererProps } from 'react-table';
+
+export type DefaultCellProps<
+  T extends Record<string, unknown>
+> = CellRendererProps<T> & React.ComponentPropsWithoutRef<'div'>;
+
+/**
+ * Default cell.
+ * It should be passed to `cellRenderer`.
+ * In most cases you shouldn't need to pass this component.
```
Comment (veekeys) [inline src/core/Table/cells/DefaultCell.tsx:15]: Added this so users could add some native attributes on the Cell itself, right? So maybe instead of this comment, we could say that `Use this component to add native html attributes to the cell container` or something like that...

File: stories/core/Table.stories.tsx [hunk 19]
```diff
@@ -1324,3 +1327,117 @@ Full.args = {
...
+    (columnId: string, value: string, rowData: TableStoryDataType) => {
+      action('onCellEdit')({ columnId, value, rowData });
+      setData((oldData) => {
+        const newData = [...oldData];
+        const index = oldData.indexOf(rowData);
+        const newObject = { ...newData[index] };
+        newObject[columnId as keyof TableStoryDataType] = value;
+        newData[index] = newObject;
+        return newData;
```
Comment (veekeys) [inline stories/core/Table.stories.tsx:1358]: This is weird... Interface expect cellEdit with void return, but here you return something.. does it have influence?

File: src/core/Table/EditableCell.tsx (hunk 1 above)
Comment (bentleyvk) [reply src/core/Table/EditableCell.tsx]: Changed to just params

File: src/core/Table/Table.test.tsx (hunk 17 above)
Comment (bentleyvk) [reply src/core/Table/Table.test.tsx:1533]: image

Comment (veekeys) [reply src/core/Table/Table.test.tsx:1533]: oh...
Could test both cases then, but minor, I guess..

File: stories/core/Table.stories.tsx (hunk 19 above)
Comment (bentleyvk) [reply stories/core/Table.stories.tsx:1358]: Huh? It is a callback inside `setData`.

Comment (veekeys) [reply stories/core/Table.stories.tsx:1358]: oh crap... yeah, alignment didnt help :D

File: src/core/Table/TableCell.tsx (hunk 16 above)
Comment (bentleyvk) [reply src/core/Table/TableCell.tsx:22]: I wanted to avoid confusion between this component and `DefaultCell`.

File: src/core/Table/TableCell.tsx (hunk 15 above)
Comment (bentleyvk) [reply src/core/Table/TableCell.tsx]: Added

File: src/core/Table/Table.test.tsx (hunk 17 above)
Comment (bentleyvk) [reply src/core/Table/Table.test.tsx]: Added another test with edge cases.

File: src/core/Table/cells/DefaultCell.tsx (hunk 18 above)
Comment (bentleyvk) [reply src/core/Table/cells/DefaultCell.tsx]: Changed.

File: src/core/Table/cells/EditableCell.tsx [hunk 20]
```diff
@@ -41,39 +41,59 @@ export type EditableCellProps<
...
+      onKeyDown={(e) => {
+        // Prevents from adding HTML elements (div, br) inside a cell on Enter press
+        if (e.key === 'Enter') {
+          e.preventDefault();
+        }
+        props.onKeyDown?.(e);
+      }}
+      onPaste={(e) => {
+        e.preventDefault();
+        document.execCommand(
```
Comment (veekeys) [inline src/core/Table/cells/EditableCell.tsx:85]: image

File: stories/core/Table.stories.tsx [hunk 21]
```diff
@@ -1324,3 +1327,120 @@ Full.args = {
...
+      });
+    },
+    [],
+  );
+
+  const cellRenderer = useCallback(
+    (props: CellRendererProps<TableStoryDataType>) => (
+      <>
+        {!isRowDisabled(props.cellProps.row.original) &&
+        props.cellProps.value !== 'Fetching...' ? (
```
Comment (veekeys) [inline stories/core/Table.stories.tsx:1368]: why cursor changes on `fetching` cell even though its not editable?

Comment (veekeys) [reply stories/core/Table.stories.tsx:1368]: ahh, seems it was like that before everywhere...nvm

File: src/core/Table/cells/EditableCell.tsx (hunk 20 above)
Comment (bentleyvk) [reply src/core/Table/cells/EditableCell.tsx:85]: I know... but there isn't any alternatives in new APIs.

**Review by veekeys — APPROVED** [review]
